File: backend/app/schemas/spatial.py
"""Pydantic response contracts for the public spatial API."""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class GeoJsonGeometry(BaseModel):
    """A GeoJSON geometry returned by PostGIS through PostgREST."""

    type: str
    coordinates: Any


class ParcelResponse(BaseModel):
    id: UUID
    base_ulpin: str
    geometry: GeoJsonGeometry
    state_code: str
    district_code: str
    created_at: str


class BuildingResponse(BaseModel):
    id: UUID | str
    parcel_id: UUID | str | None = None
    base_ulpin: str
    building_code: str
    building_type: str
    structure_category: str | None = None
    name: str | None = None
    house_no: str | None = None
    complex_name: str | None = None
    complex_type: str | None = None
    stories_count: int | None = None
    height_meters: float | None = None
    floors_completed: int | None = None
    floors_planned: int | None = None
    footprint: GeoJsonGeometry
    center: list[float] | None = None
    bounds: list[float] | None = None
    ai_spatial_hash: str | None = None
    postal_address: str | None = None
    total_units_count: int | None = None
    created_at: str | None = None


class FloorResponse(BaseModel):
    id: UUID | str
    building_id: UUID | str
    floor_code: str
    floor_number: int
    is_underground: bool = False
    footprint: GeoJsonGeometry
    height_meters: float = 3.5


class UnitResponse(BaseModel):
    id: UUID | str
    floor_id: UUID | str
    unit_code: str
    full_ulpin: str
    unit_number: str | None = None
    footprint: GeoJsonGeometry
    unit_type: str
    elevation_meters: list[float] | None = None
    volume_m3: float | None = None
    ai_morton_code: str | None = None
    spatial_verification_hash: str | None = None


class UnitDetailResponse(UnitResponse):
    base_ulpin: str
    building_code: str
    floor_code: str
    assembled_ulpin: str
    building_name: str | None = None
    house_no: str | None = None
    complex_name: str | None = None
    structure_category: str | None = None
    stories_count: int | None = None
    postal_address: str | None = None


class UndergroundInfraResponse(BaseModel):
    id: UUID | str
    parcel_id: UUID | str
    base_ulpin: str
    infra_type: Literal["drainage", "metro_tunnel", "metro_station"]
    path: GeoJsonGeometry
    full_ulpin: str
    depth_meters: float


class SearchRecord(BaseModel):
    record_type: Literal["parcel", "building", "floor", "unit"]
    id: UUID | str
    base_ulpin: str
    building_code: str | None = None
    floor_code: str | None = None
    unit_code: str | None = None
    full_ulpin: str
    name: str | None = None
    house_no: str | None = None
    geometry: GeoJsonGeometry


class SearchResponse(BaseModel):
    query: str
    records: list[SearchRecord] = Field(default_factory=list)

File: frontend/src/types/spatial.ts
/** TypeScript contracts mirroring the Phase 4 FastAPI response schemas. */

export interface GeoJsonGeometry {
  type: string
  coordinates: unknown
}

export interface Parcel {
  id: string
  base_ulpin: string
  geometry: GeoJsonGeometry
  state_code: string
  district_code: string
  created_at: string
}

export type BuildingType = 'apartment' | 'house' | 'half_built' | 'school' | 'commercial'

export interface RawCatalogUnit {
  unit_code: string
  full_ulpin: string
  unit_number?: string | null
  unit_type: string
  elevation_meters?: [number, number] | null
  volume_m3?: number | null
  ai_morton_code?: string | null
  spatial_verification_hash?: string | null
}

export interface RawCatalogFloor {
  id?: string
  floor_code: string
  floor_number: number
  height_meters?: number
  units?: RawCatalogUnit[]
}

export interface Building {
  id: string
  parcel_id?: string | null
  base_ulpin: string
  building_code: string
  building_type: string
  structure_category?: string | null
  name?: string | null
  house_no?: string | null
  complex_name?: string | null
  complex_type?: string | null
  stories_count?: number | null
  height_meters?: number | null
  floors_completed?: number | null
  floors_planned?: number | null
  footprint: GeoJsonGeometry
  center?: [number, number] | null
  bounds?: [number, number, number, number] | null
  ai_spatial_hash?: string | null
  postal_address?: string | null
  total_units_count?: number | null
  created_at?: string | null
  floors?: RawCatalogFloor[]
}

export interface Floor {
  id: string
  building_id: string
  floor_code: string
  floor_number: number
  is_underground: boolean
  footprint: GeoJsonGeometry
  height_meters: number
}

export type UnitType = 'residential' | 'commercial' | 'common_area' | string

export interface Unit {
  id: string
  floor_id: string
  unit_code: string
  full_ulpin: string
  unit_number?: string | null
  footprint: GeoJsonGeometry
  unit_type: UnitType
  elevation_meters?: [number, number] | null
  volume_m3?: number | null
  ai_morton_code?: string | null
  spatial_verification_hash?: string | null
}

export interface UlpinBreakdown {
  base_ulpin: string
  building_code: string
  floor_code: string
  unit_code: string
  assembled_ulpin: string
}

export interface UnitDetail extends Unit, UlpinBreakdown {
  building_name?: string | null
  house_no?: string | null
  complex_name?: string | null
  structure_category?: string | null
  stories_count?: number | null
  postal_address?: string | null
}

export type InfraType = 'drainage' | 'metro_tunnel' | 'metro_station'

export interface UndergroundInfra {
  id: string
  parcel_id?: string | null
  base_ulpin: string
  infra_type: InfraType
  path: GeoJsonGeometry
  full_ulpin: string
  depth_meters: number
  /** Three.js scene-coordinate waypoints [x, y, z] — negative y = underground. */
  waypoints?: [number, number, number][]
  diameter_m?: number
  material?: string
  /** Assembled hierarchical ULPIN (only present for metro_station nodes). */
  assembled_ulpin?: string
  segment_name?: string
}

export type SearchRecordType = 'parcel' | 'building' | 'floor' | 'unit'

export interface SearchRecord {
  record_type: SearchRecordType
  id: string
  base_ulpin: string
  building_code: string | null
  floor_code: string | null
  unit_code: string | null
  full_ulpin: string
  name?: string | null
  house_no?: string | null
  geometry: GeoJsonGeometry
}

export interface SearchResponse {
  query: string
  records: SearchRecord[]
}
